src/hotspot/share/gc/shared/partialArrayState.cpp

```diff
@@ -81,73 +62,79 @@ struct PartialArrayStateAllocator::Impl::FreeListEntry {
   NONCOPYABLE(FreeListEntry);
 };
 
-PartialArrayStateAllocator::Impl::Impl(uint num_workers)
-  : _arenas(NEW_C_HEAP_ARRAY(Arena, num_workers, mtGC)),
-    _free_lists(NEW_C_HEAP_ARRAY(FreeListEntry*, num_workers, mtGC)),
-    _num_workers(num_workers)
-{
-  for (uint i = 0; i < _num_workers; ++i) {
-    ::new (&_arenas[i]) Arena(mtGC);
-    _free_lists[i] = nullptr;
-  }
-}
+PartialArrayStateAllocator::PartialArrayStateAllocator(PartialArrayStateManager* manager)
+  : _manager(manager),
+    _free_list(),
+    _arena(manager->register_allocator())
+{}
 
-PartialArrayStateAllocator::Impl::~Impl() {
-  // We don't need to clean up the free lists.  Deallocating the entries
+PartialArrayStateAllocator::~PartialArrayStateAllocator() {
+  // We don't need to clean up the free list.  Deallocating the entries
   // does nothing, since we're using arena allocation.  Instead, leave it
-  // to the arena destructor to release the memory.
-  FREE_C_HEAP_ARRAY(FreeListEntry*, _free_lists);
-  for (uint i = 0; i < _num_workers; ++i) {
-    _arenas[i].~Arena();
-  }
-  FREE_C_HEAP_ARRAY(Arena*, _arenas);
+  // to the manager to release the memory.
+  // Inform the manager that an allocator is no longer in use.
+  _manager->release_allocator();
 }
 
-PartialArrayState* PartialArrayStateAllocator::Impl::allocate(uint worker_id,
-                                                              oop src, oop dst,
-                                                              size_t index,
-                                                              size_t length,
-                                                              size_t initial_refcount) {
+PartialArrayState* PartialArrayStateAllocator::allocate(oop src, oop dst,
+                                                        size_t index,
+                                                        size_t length,
+                                                        size_t initial_refcount) {
   void* p;
-  FreeListEntry* head = _free_lists[worker_id];
+  FreeListEntry* head = _free_list;
   if (head == nullptr) {
-    p = NEW_ARENA_OBJ(&_arenas[worker_id], PartialArrayState);
+    p = NEW_ARENA_OBJ(_arena, PartialArrayState);
   } else {
-    _free_lists[worker_id] = head->_next;
+    _free_list = head->_next;
     head->~FreeListEntry();
     p = head;
   }
   return ::new (p) PartialArrayState(src, dst, index, length, initial_refcount);
 }
 
-void PartialArrayStateAllocator::Impl::release(uint worker_id, PartialArrayState* state) {
+void PartialArrayStateAllocator::release(PartialArrayState* state) {
   size_t refcount = Atomic::sub(&state->_refcount, size_t(1), memory_order_release);
```

Comment:
Could you explain why `release` order is needed here?

Reply:
This is part of the usual reference counting dance. Except, where did the
acquire disappear to? There should be an acquire on the refcount == 0 branch!
Looks like I accidentally deleted it. Sigh. Not too surprisingly, lots of
tests were run without noticing that.

Reply:
Make sense. My next question is that, if `PartialArrayState` is ever crossed thread boundaries, it is through job stealing via task queues. Can we depend on barriers of task queue to ensure the memory safety? because we don't need any additional barriers for objects popped/stole from task queues in other places.

Reply:
Yes, it's needed.  The purpose of this release/acquire pair is to ensure there
is a happens-before chain between the use of a State and it's cleanup/reuse.
Transfers through the taskqueue don't help with that.

In other places, we have operations on one side of the taskqueue that need to
be ordered wrto operations on the other side of the taskqueue.  We don't have
that here.

Consider two threads which have both obtained access to a State.  (At least
one of them must have obtained it via stealing from another thread.)  Assume
these are the last two references to the State (it's refcount == 2), and no
further tasks for it will be needed.  These two threads will use the State
(getting source/destination, claiming a chunk), and then release the State,
decrementing its refcount.  So one of them will decrement to 0.  We need to
ensure the cleanup that follows that can't corrupt the use by the accesses
made by the other thread, by ensuring those accesses happen-before the
cleanup.  There is no intervening taskqueue manipulation in this scenario.
The operations we need to order are all on the same side of the taskqueue.

Reply:
I see.